addon/addon/models/sprite-snapshot-node-builder.ts

```diff
@@ -0,0 +1,344 @@
+          );
+        }
+
+        this.contextToNode.set(context, spriteSnapshotNode);
+      } else {
+        // We already decided what contexts we're going to use for this render,
+        // so we can mark new contexts for the next run.
+        context.isInitialRenderCompleted = true;
+      }
+    }
+  }
+
+  classifySprites(
+    freshlyAdded: Set<SpriteModifier>,
+    freshlyRemoved: Set<SpriteModifier>,
+    freshlyChanged: Set<SpriteModifier>
+  ) {
+    let insertedSpritesArr = [...freshlyAdded];
+    let removedSpritesArr = [...freshlyRemoved];
+
+    let classifiedInsertedSpriteModifiers = new Set(insertedSpritesArr);
+    let classifiedRemovedSpriteModifiers = new Set(removedSpritesArr);
+    let classifiedKeptSpriteModifiers = new Set([...freshlyChanged]);
+
+    let spriteModifiers: Set<SpriteModifier> = new Set();
+    let spriteModifierToSpriteMap = new WeakMap<SpriteModifier, Sprite>();
+    let spriteModifierToCounterpartModifierMap = new WeakMap<
+      SpriteModifier,
+      SpriteModifier
+    >();
+
+    // Collect intersecting sprite identifiers
+    let insertedIds = insertedSpritesArr.map(
+      (s) => new SpriteIdentifier(s.id, s.role)
+    );
+    let removedIds = removedSpritesArr.map(
+      (s) => new SpriteIdentifier(s.id, s.role)
+    );
+    let intersectingIds = insertedIds.filter(
+      (identifier) => !!removedIds.find((o) => o.equals(identifier))
+    );
+
+    // Classify non-natural KeptSprites
+    for (let intersectingId of intersectingIds) {
+      let insertedSpriteModifier = insertedSpritesArr.find((s) =>
+        new SpriteIdentifier(s.id, s.role).equals(intersectingId)
+      );
+      let removedSpriteModifiers = removedSpritesArr.filter((s) =>
+        new SpriteIdentifier(s.id, s.role).equals(intersectingId)
+      );
+
+      assert(
+        'Intersection check should always result in removedSpriteModifier and insertedSpriteModifier being found',
+        !(!insertedSpriteModifier || removedSpriteModifiers.length === 0)
+      );
+      assert(
+        'Multiple matching removedSpriteModifiers found',
+        removedSpriteModifiers.length < 2
+      );
+
+      let removedSpriteModifier = removedSpriteModifiers[0] as SpriteModifier;
+      classifiedKeptSpriteModifiers.add(insertedSpriteModifier);
+      classifiedInsertedSpriteModifiers.delete(insertedSpriteModifier);
+      classifiedRemovedSpriteModifiers.delete(removedSpriteModifier);
+
+      let keptSprite = new Sprite(
+        insertedSpriteModifier.element as HTMLElement,
+        insertedSpriteModifier.id,
+        insertedSpriteModifier.role,
+        SpriteType.Kept
+      );
+      keptSprite.counterpart = new Sprite(
+        removedSpriteModifier.element as HTMLElement,
+        removedSpriteModifier.id,
+        removedSpriteModifier.role,
+        SpriteType.Removed
+      );
+      spriteModifiers.add(insertedSpriteModifier);
+      spriteModifierToSpriteMap.set(insertedSpriteModifier, keptSprite);
```

Comment:
After changing the inserted sprite into a kept sprite between contexts, overflow can become an issue if the inner context is managing the new kept sprite. 

https://user-images.githubusercontent.com/39579264/174632115-7984c4e3-effc-4c7a-af5a-67cb5567b31b.mov

Reply:
This is interesting, we talked about taking the first shared ancestor context for kept sprites I think? Must not be working yet if this is happening 😅